core/server/common/src/main/java/alluxio/master/BackupManager.java

```diff
@@ -86,15 +90,48 @@ public void initFromBackup(InputStream is) throws IOException {
     try (GzipCompressorInputStream gzIn = new GzipCompressorInputStream(is);
          JournalEntryStreamReader reader = new JournalEntryStreamReader(gzIn)) {
       List<Master> masters = mRegistry.getServers();
-      JournalEntry entry;
       Map<String, Master> mastersByName = Maps.uniqueIndex(masters, Master::getName);
-      while ((entry = reader.readEntry()) != null) {
-        String masterName = JournalEntryAssociation.getMasterForEntry(entry);
-        Master master = mastersByName.get(masterName);
-        master.processJournalEntry(entry);
-        try (JournalContext jc = master.createJournalContext()) {
-          jc.append(entry);
-          count++;
+      // Create buffer to avoid flushing per-entry.
+      int bufferLimit = Configuration.getInt(PropertyKey.MASTER_BACKUP_ENTRY_BUFFER_COUNT);
+      List<JournalEntry> entryBuffer = new ArrayList<>(bufferLimit);
+      // Apply entries from input stream.
+      JournalEntry entry;
+      while (true) {
+        // Read next entry.
+        entry = reader.readEntry();
+        // Buffer next entry.
+        if (entry != null) {
+          entryBuffer.add(entry);
+        }
+        // Process the buffer if it is full or the input stream has no more entries.
+        if (entry == null || entryBuffer.size() >= bufferLimit) {
+          // Used to keep track of journal contexts per master.
+          Map<String, JournalContext> contextMap = new HashMap<>();
+          // Process each buffered entry.
+          for (JournalEntry bufferedEntry : entryBuffer) {
+            String masterName = JournalEntryAssociation.getMasterForEntry(bufferedEntry);
+            Master master = mastersByName.get(masterName);
+            // Apply entry to master.
+            master.processJournalEntry(bufferedEntry);
+            // Create journal context if master is seens the first time.
+            if (!contextMap.containsKey(masterName)) {
+              contextMap.put(masterName, master.createJournalContext());
+            }
+            // Apply entry to journal.
+            contextMap.get(masterName).append(bufferedEntry);
```

Comment:
Can we assume all of these operations succeed? Otherwise it might be risky to apply the change to internal master state before the journal?

Reply:
If applying a backup throws, then masters won't come up.